### PR TITLE
Pause transaction

### LINKED ISSUE #1315: Ability to pause transactions
Add a new pallet & signed extension that allow council to set custom filter on transactions. This can be used to temporarily disable certain transactions without runtime upgrade, which requires 1 hour waiting time.

### PR DESCRIPTION
**Pull request by wangjj9219**
close #1315

### CODE REVIEW

File: modules/transaction-pause/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+		// can not pause
+		CannotPause,
+		// invalid character encoding
+		InvalidCharacter,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(fn deposit_event)]
+	pub enum Event<T: Config> {
+		TransactionPaused(Vec<u8>, Vec<u8>),
```
Comment (xlc) [inline modules/transaction-pause/src/lib.rs:61]: comments

File: modules/transaction-pause/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+		#[pallet::weight(T::WeightInfo::unpause_transaction())]
+		#[transactional]
+		pub fn unpause_transaction(
+			origin: OriginFor<T>,
+			pallet_name: Vec<u8>,
+			function_name: Vec<u8>,
+		) -> DispatchResult {
+			T::UpdateOrigin::ensure_origin(origin)?;
+			if PausedTransactions::<T>::take((pallet_name.clone(), function_name.clone())).is_some() {
```
Comment (xlc) [inline modules/transaction-pause/src/lib.rs:112]:
```suggestion
			if PausedTransactions::<T>::take((&pallet_name, &function_name)).is_some() {
```

File: modules/transaction-pause/src/tests.rs [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+#[test]
+fn pause_transaction_work() {
+	ExtBuilder::default().build().execute_with(|| {
+		System::set_block_number(1);
+
+		assert_noop!(
+			TransactionPause::pause_transaction(
+				Origin::signed(5),
+				"Balances".as_bytes().to_vec(),
```
Comment (xlc) [inline modules/transaction-pause/src/tests.rs:41]:
```suggestion
				b"Balances".to_vec(),
```

File: modules/transaction-pause/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(T::WeightInfo::pause_transaction())]
+		#[transactional]
+		pub fn pause_transaction(origin: OriginFor<T>, pallet_name: Vec<u8>, function_name: Vec<u8>) -> DispatchResult {
+			T::UpdateOrigin::ensure_origin(origin)?;
+
+			// because the pallet name of this module in runtime is undetermined here,
+			// not allowed to pause calls have same names with these calls to ensure safe
+			// event if maybe there are calls with the same name in other pallet.
+			let function_name_string =
```
Comment (xlc) [inline modules/transaction-pause/src/lib.rs:88]: it should be possible to get pallet name using `PalletInfoAccess`

https://github.com/paritytech/substrate/blob/71e1c59e7e901882487ee516cf2faa02aed74da5/frame/support/test/tests/pallet.rs#L1265

**Review by xlc — APPROVED** [review]

File: modules/transaction-pause/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+		/// The origin which may set filter.
+		type UpdateOrigin: EnsureOrigin<Self::Origin>;
+
+		/// Weight information for the extrinsics in this module.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		// can not pause
```
Comment (shaunxw) [inline modules/transaction-pause/src/lib.rs:56]:
```suggestion
		/// can not pause
```

File: modules/transaction-pause/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+		CannotPause,
+		// invalid character encoding
```
Comment (shaunxw) [inline modules/transaction-pause/src/lib.rs:58]:
```suggestion
		/// invalid character encoding
```

**Review by shaunxw — APPROVED** [review]
Will be good to have some integration tests

**Comment by ermalkaleci** [comment]
/bench runtime custom-karura --chain=karura-dev --steps=50 --repeat=20 --pallet=module_transaction_pause --extrinsic=* --execution=wasm --wasm-execution=compiled --heap-pages=4096 --template=./templates/runtime-weight-template.hbs --output=./runtime/karura/src/weights/

**Comment by wangjj9219** [comment]
/bench runtime custom-karura --chain=karura-dev --steps=50 --repeat=20 --pallet=module_transaction_pause --extrinsic=* --execution=wasm --wasm-execution=compiled --heap-pages=4096 --template=./templates/runtime-weight-template.hbs --output=./runtime/karura/src/weights/

**Review by xlc — APPROVED** [review]

**Review by shaunxw — APPROVED** [review]

**Pull request merged**